---
title: "CellChatAnalysis"
author: "Daniel Lee"
date: "2023-08-01"
output: pdf_document
---

## Overall analysis
```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE,  
                      warning = FALSE, message = FALSE, 
                      fig.align = "center",
                      R.options = list(max.print=100))
```

```{r}
getwd()

# running renv
# install.packages("renv")
library(renv)
# renv::init()
# renv::snapshot()
# renv::status()
# renv::clean()
# renv::restore()
# renv::history()
```

```{r}
# install.packages("devtools")

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.17")

# BiocManager::install("NMF")
# BiocManager::install("BiocGenerics")
# install.packages("car")
# BiocManager::install("Biobase")
# BiocManager::install("ComplexHeatmap")
# BiocManager::install("Seurat")

# svglite, expm, RSpectra, ggpubr, car, gert, mgcv, BiocParallel, BiocNeighbors, clue manually downloaded from CRAN & BiocManager and moved to renv project directory
# devtools::install_github("sqjin/CellChat")

library(CellChat)
library(Seurat)

```

```{r}
# Load your data
setwd("/Volumes/biologos/Bioinformatics/230731_scRNA_CellChat/Data")
lymphoid <- readRDS("lymphoid_srt_int.rds")
myeloid <- readRDS("myeloid_srt_int.rds")
# Combining the two datasets into a single Seurat object
combined_data <- merge(x = lymphoid, y = myeloid)
```

```{r}
# Define cell types to keep
cell_types_to_keep <- c("CD3+CD8+ T cells II", "CD14+CD68+ Macrophages II", "CD3+CD4+ T cells I", "CD3+CD8+ T Cells III", "CD3+CD8+ T cells I","CD3+CD4+ T Cells II","CD14+CD68+ Macrophages I","CLEC4+ plasmacytoid DCs")

# Subset data based on metadata
subset_seurat <- subset(x = data, subset = cell_type %in% cell_types_to_keep)

# Drop unused factor levels in cell_type
subset_seurat@meta.data$cell_type <- droplevels(subset_seurat@meta.data$cell_type)

# Check unique cell types in the new Seurat object
unique(subset_seurat@meta.data$cell_type)

# Create a new CellChat object with the subsetted data
cellchat <- createCellChat(object = subset_seurat, meta = subset_seurat@meta.data, group.by = "cell_type")
```

```{r}
# Set the ligand-receptor interaction database
CellChatDB.use <- CellChatDB.human

# Using a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB.human, search = "Secreted Signaling")

# Add the Secreted Signaling database in the CellChat object
cellchat@DB <- CellChatDB.use

```

```{r}
# Pre-process the expression data
cellchat <- subsetData(cellchat)
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)

```

```{r}
# Project gene expression data onto protein-protein interaction (PPI)
cellchat <- projectData(cellchat, PPI.human)
```

```{r}
# Compute communication probabilities again
cellchat <- computeCommunProb(cellchat, raw.use = FALSE)

```

```{r}
# Filter out the cell-cell communication if there are only few number of cells 
# in certain cell groups
cellchat <- filterCommunication(cellchat, min.cells = 10)

```

```{r}
# Infer the cell-cell communication at a signaling pathway level
cellchat <- computeCommunProbPathway(cellchat)

```

```{r}
# Calculate the aggregated cell-cell communication network
cellchat <- aggregateNet(cellchat)
cellchat@net$count
cellchat@net$weight

```

```{r}
# Create your plots
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1, 2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, 
   weight.scale = T, label.edge= T, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, 
   weight.scale = T, label.edge= T, title.name = "Interaction weights/strength")
```

```{r}
str(cellchat)

# Save the CellChat object for future reference
# saveRDS(cellchat, file = "Myeloid_Lymphatic_snRNA_snATAC_analysed.rds")

```

```{r}
# table(cellchat@meta$patient_region_id)
table(cellchat@meta$orig.ident)
```
## Femoral sample analysis
```{r}
# Define femoral v.s. carotid
carotid_ids <- c("carotid 1 & 2", "carotid 3", "carotid 4")
femoral_ids <- setdiff(unique(subset_data@meta.data$orig.ident), carotid_ids)

# Subset the combined data for each group
carotid_data <- subset_data[, subset_data@meta.data$orig.ident %in% carotid_ids]

femoral_data <- subset_data[, subset_data@meta.data$orig.ident %in% femoral_ids]
```

```{r}
# Control group
cellchat_carotid <- createCellChat(object = carotid_data, meta = carotid_data@meta.data, group.by = "cell_type")

# Set the ligand-receptor interaction database
CellChatDB.use <- CellChatDB.human

# Using a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB.human, search = "Secreted Signaling")

# Add the Secreted Signaling database in the CellChat object
cellchat_carotid@DB <- CellChatDB.use

```

```{r}
# Pre-process the expression data
cellchat_carotid <- subsetData(cellchat_carotid)
cellchat_carotid <- identifyOverExpressedGenes(cellchat_carotid)
cellchat_carotid <- identifyOverExpressedInteractions(cellchat_carotid)

```

```{r}
# Project gene expression data onto protein-protein interaction (PPI)
cellchat_carotid <- projectData(cellchat_carotid, PPI.human)
```

```{r}
# Compute communication probabilities again
cellchat_carotid <- computeCommunProb(cellchat_carotid, raw.use = FALSE)

```

```{r}
# Filter out the cell-cell communication if there are only few number of cells 
# in certain cell groups
cellchat_carotid <- filterCommunication(cellchat_carotid, min.cells = 10)

```

```{r}
# Infer the cell-cell communication at a signaling pathway level
cellchat_carotid <- computeCommunProbPathway(cellchat_carotid)

```

```{r}
# Calculate the aggregated cell-cell communication network
cellchat_carotid <- aggregateNet(cellchat_carotid)
cellchat_carotid@net$count
cellchat_carotid@net$weight

```

## Carotid analysis
```{r}
# Diseased group
cellchat_diseased <- createCellChat(object = diseased_data, meta = diseased_data@meta.data, group.by = "annotation")

# Set the ligand-receptor interaction database
CellChatDB.use <- CellChatDB.human

# Using a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB.human, search = "Secreted Signaling")

# Add the Secreted Signaling database in the CellChat object
cellchat_diseased@DB <- CellChatDB.use

```

```{r}
# Pre-process the expression data
cellchat_diseased <- subsetData(cellchat_diseased)
cellchat_diseased <- identifyOverExpressedGenes(cellchat_diseased)
cellchat_diseased <- identifyOverExpressedInteractions(cellchat_diseased)

```

```{r}
# Project gene expression data onto protein-protein interaction (PPI)
cellchat_diseased <- projectData(cellchat_diseased, PPI.human)
```

```{r}
# Compute communication probabilities again
cellchat_diseased <- computeCommunProb(cellchat_diseased, raw.use = FALSE)

```

```{r}
# Filter out the cell-cell communication if there are only few number of cells 
# in certain cell groups
cellchat_diseased <- filterCommunication(cellchat_diseased, min.cells = 10)


```

```{r}
# Infer the cell-cell communication at a signaling pathway level
cellchat_diseased <- computeCommunProbPathway(cellchat_diseased)

```

```{r}
# Calculate the aggregated cell-cell communication network
cellchat_diseased <- aggregateNet(cellchat_diseased)
cellchat_diseased@net$count
cellchat_diseased@net$weight

```

## Network plot visualization of healthy v.s. M.I. populations
```{r}
# Calculate group size for each cellchat object
groupSize_control <- as.numeric(table(cellchat_carotid@idents))
groupSize_diseased <- as.numeric(table(cellchat_diseased@idents))

# Initialize the plot with a layout of 2x2 so that each condition can have two plots (one for interaction count and one for interaction weight)
par(mfrow = c(2, 2), xpd=TRUE)

png("NetworkPlot_NofInteractions_Control.png", width=5, height=5, units="in", res=600)
netVisual_circle(cellchat_carotid@net$count, vertex.weight = groupSize_control, 
   weight.scale = T, label.edge= T, title.name = "Control: Number of interactions")
dev.off()
png("NetworkPlot_InteractionStrength_Control.png", width=5, height=5, units="in", res=600)
netVisual_circle(cellchat_carotid@net$weight, vertex.weight = groupSize_control, 
   weight.scale = T, label.edge= F, title.name = "Control: Interaction weights/strength")
dev.off()

png("NetworkPlot_NofInteractions_MI.png", width=5, height=5, units="in", res=600)
netVisual_circle(cellchat_diseased@net$count, vertex.weight = groupSize_diseased, 
   weight.scale = T, label.edge= T, title.name = "M.I.: Number of interactions")
dev.off()
png("NetworkPlot_InteractionStrength_MI.png", width=5, height=5, units="in", res=600)
netVisual_circle(cellchat_diseased@net$weight, vertex.weight = groupSize_diseased, 
   weight.scale = T, label.edge= F, title.name = "M.I.: Interaction weights/strength")
dev.off()

```

## Pathway analysis
```{r}
png("InteractionStrength_Control.png", width=8, height=8, units="in", res=600)
cellchat_carotid1 <- netAnalysis_computeCentrality(cellchat_carotid)
netAnalysis1 <- netAnalysis_signalingRole_scatter(cellchat_carotid1)
print(netAnalysis1)
dev.off()

png("InteractionStrength_MI.png", width=8, height=8, units="in", res=600)
cellchat_diseased1 <- netAnalysis_computeCentrality(cellchat_diseased)
netAnalysis2 <- netAnalysis_signalingRole_scatter(cellchat_diseased1)
print(netAnalysis2)
dev.off()

# show all the interactions sending from MACROPHAGES to LECs
# Chord diagram
png("ChordDiagram_Control.png", width=20, height=20, units="in", res=600)
chord1 <- netVisual_chord_gene(cellchat_carotid1, sources.use = c(1,3,4,5), targets.use = 2, lab.cex = 1,show.legend=TRUE)
print(chord1)
dev.off()

png("ChordDiagram_MI.png", width=20, height=20, units="in", res=600)
chord2 <- netVisual_chord_gene(cellchat_diseased1, sources.use = c(1,3,4,5), targets.use = 2, lab.cex = 1,show.legend=TRUE)
print(chord2)
dev.off()

# Bubble plot
png("BubblePlot_Control.png", width=7, height=7, units="in", res=600)
bubble1 <- netVisual_bubble(cellchat_carotid1, sources.use = c(1,3,4,5), targets.use = 2, remove.isolate = TRUE, angle.x=45)
print(bubble1)
dev.off()

png("BubblePlot_MI.png", width=7, height=10, units="in", res=600)
bubble2 <- netVisual_bubble(cellchat_diseased1, sources.use = c(1,3,4,5), targets.use = 2, remove.isolate = FALSE, angle.x=45)
print(bubble2)
dev.off()
```


## Violin plots of healthy v.s. MI'd populations
```{r}
# install.packages("patchwork")
library(patchwork)

# Define gene list
gene_list <- c("FLT4", "KDR", "VASH1", "CCBE1", "CD36", "SYK", "MERTK", "AXL", "VEGFC", "IL1B", "IL10", "TNF")

# Normalize and scale control data
control_data <- NormalizeData(control_data)
control_data <- ScaleData(control_data, features = gene_list)

# Normalize and scale diseased data
diseased_data <- NormalizeData(diseased_data)
diseased_data <- ScaleData(diseased_data, features = gene_list)

# Generate violin plots for control cells
p_list <- list()
for (gene in gene_list) {
  p_list[[gene]] <- print(VlnPlot(control_data, features = gene, group.by = "annotation"))
}

# Combine all plots
p_combined <- wrap_plots(p_list, ncol = 4)

# Print the combined plot
print(p_combined)

# Generate violin plots for diseased cells
p_list2 <- list()
for (gene in gene_list) {
  p_list2[[gene]] <- print(VlnPlot(diseased_data, features = gene, group.by = "annotation"))
}

# Combine all plots
p_combined2 <- wrap_plots(p_list2, ncol = 4)

# Print the combined plot
print(p_combined2)

# Save the combined plot
ggsave("Control_Violin_Plots.png", p_combined, width=45, height=35,dpi=600)
ggsave("MI_Violin_Plots.png", p_combined2, width=45, height=35,dpi=600)

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r}

```

```{r, echo=F}
## DO NOT DELETE THIS BLOCK!
Sys.info()
sessionInfo()
```